datafusion/physical-plan/src/sorts/sort.rs

```diff
@@ -673,29 +676,211 @@ impl ExternalSorter {
+                // Use `lexsort_to_indices` to get global row indices in sorted order (as if all batches were concatenated)
+
+                let indices = lexsort_to_indices(&sort_columns, None)?;
+
+                // ===== Phase 3: Reorder each column using the global sorted indices =====
+                let batch_indices: Vec<(usize, usize)> = current_batches
+                    .iter()
+                    .enumerate()
+                    .map(|(batch_id, batch)| {
+                        (0..batch.num_rows()).map(move |i| (batch_id, i))
+                    })
+                    .flatten()
+                    .collect();
+
+                // For each column:
+                // 1. Concatenate all batch arrays for this column (in the same order as assumed by `lexsort_to_indices`)
+                // 2. Use Arrow's `take` function to reorder the column by sorted indices
+                let interleave_indices: Vec<(usize, usize)> = indices
+                    .values()
+                    .iter()
+                    .map(|x| batch_indices[*x as usize])
+                    .collect();
+                // Build a RecordBatch from the sorted columns
+
+                let batches: Vec<&RecordBatch> = current_batches.iter().collect();
+
+                let sorted_batch =
+                    interleave_record_batch(batches.as_ref(), &interleave_indices)?;
+
+                current_batches.clear();
+
+                // Update memory reservation.
+                self.reservation.try_shrink(current_size)?;
+                let merged_size = get_reserved_byte_for_record_batch(&sorted_batch);
+                self.reservation.try_grow(merged_size)?;
+
+                merged_batches.push(sorted_batch);
+                current_size = 0;
+            }
+
+            current_batches.push(batch);
+            current_size += batch_size;
+        }
+
+        // Merge any remaining batches after the loop.
+        if !current_batches.is_empty() {
+            // ===== Phase 1: Build global sort columns for each sort expression =====
+            // For each sort expression, evaluate and collect the corresponding sort column from each in-memory batch
+            // Here, `self.expr` is a list of sort expressions, each providing `evaluate_to_sort_column()`,
+            // which returns an ArrayRef (in `.values`) and sort options (`options`)
+
+            /// ```text
+            /// columns_by_expr for example:
+            /// ├── expr_0 ──┬── ArrayRef_0_0 (from batch_0)
+            /// │            ├── ArrayRef_0_1 (from batch_1)
+            /// │            └── ArrayRef_0_2 (from batch_2)
+            /// │
+            /// └── expr_1 ──┬── ArrayRef_1_0 (from batch_0)
+            ///              ├── ArrayRef_1_1 (from batch_1)
+            ///              └── ArrayRef_1_2 (from batch_2)
+            /// ```
+            let mut columns_by_expr: Vec<Vec<ArrayRef>> = self
```

Comment:
I am sure this can be moved to a function that sorts a list of batches (as the code is the same / almost the same)